pkg/genericapiserver/destinations.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericapiserver
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer/versioning"
+	"k8s.io/kubernetes/pkg/storage"
+	etcdstorage "k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/golang/glog"
+)
+
+// StorageDestinations is takes a GroupResource and returns back its storage interface.  This result includes:
+// 1. Merged etcd config, including: auth, server locations, prefixes
+// 2. Resource encodings for storage: group,version,kind to store as
+// 3. Cohabitating default: some resources like hpa are exposed through multiple APIs.  They must agree on 1 and 2
+type StorageDestinations struct {
+	// DefaultEtcdConfig describes how to connect to etcd in general.  It's authentication information will be used for
+	// every storage.Interface returned.
+	DefaultEtcdConfig etcdstorage.EtcdConfig
+
+	// EtcdLocationOverrides contains the list of "special" locations that are used for particular GroupResources
+	// these are merged on top of the default DefaultEtcdConfig when requesting the storage.Interface for a given GroupResource
+	EtcdLocationOverrides map[unversioned.GroupResource][]string
+
+	// EtcdPrefixOverrides contains the list of "special" prefixes for a GroupResource.  Resource=* means for the entire group
```

Comment:
> Is this new? I really don't want to allow people to use different etcd prefixes for different resources?

No, not new.  In fact, this is more constrained than before where the API made it possible to have completely unrelated storage mechanisms mixed together.  I've actually constrained the API to more closely match the intended flexibility as expose through the flags.